### PR TITLE
[Doc] Add doc about struct arguments

### PR DESCRIPTION
**Pull request by lin-hitonami**
Issue: #

### <samp>🤖 Generated by Copilot at adf95e4</samp>

This pull request improves the documentation of `ti.kernel` and `ti.func` by using the new `ti.types` module and showing how to work with structs and matrices.

### <samp>🤖 Generated by Copilot at adf95e4</samp>

*  Update the argument and return value types of kernels and functions to use the `ti.types` module ([link](https://github.com/taichi-dev/taichi/pull/7959/files?diff=unified&w=0#diff-305d66d1c231266cb14a0d2b166363840f782874eb6c5b3d5d4e15e08f3cb04cL96-R98), [link](https://github.com/taichi-dev/taichi/pull/7959/files?diff=unified&w=0#diff-305d66d1c231266cb14a0d2b166363840f782874eb6c5b3d5d4e15e08f3cb04cL138-R158), [link](https://github.com/taichi-dev/taichi/pull/7959/files?diff=unified&w=0#diff-305d66d1c231266cb14a0d2b166363840f782874eb6c5b3d5d4e15e08f3cb04cL279-R298), [link](https://github.com/taichi-dev/taichi/pull/7959/files?diff=unified&w=0#diff-305d66d1c231266cb14a0d2b166363840f782874eb6c5b3d5d4e15e08f3cb04cL287-R306), [link](https://github.com/taichi-dev/taichi/pull/7959/files?diff=unified&w=0#diff-305d66d1c231266cb14a0d2b166363840f782874eb6c5b3d5d4e15e08f3cb04cL300-R319))
* Add an example of passing a struct with a matrix to a kernel by value in `docs/lang/articles/kernels/kernel_function.md` ([link](https://github.com/taichi-dev/taichi/pull/7959/files?diff=unified&w=0#diff-305d66d1c231266cb14a0d2b166363840f782874eb6c5b3d5d4e15e08f3cb04cR112-R130))

### CODE REVIEW

File: docs/lang/articles/kernels/kernel_function.md [hunk 1]
```diff
@@ -109,6 +109,25 @@ def my_kernel(x: int, y: float):
 my_kernel(1, 1.0)  # Prints 2.0
 ```
 
+Here is another example of passing a nested struct argument with a matrix to `my_kernel()` by value:
```
Comment (neozhaoliang) [inline docs/lang/articles/kernels/kernel_function.md:112]:
```suggestion
Here is another example of passing a nested struct argument with a matrix to a kernel by value:
```
The name of the kernel below isn't `my_kernel`?

File: docs/lang/articles/kernels/kernel_function.md [hunk 2]
```diff
@@ -109,6 +109,25 @@ def my_kernel(x: int, y: float):
...
+def kernel_with_nested_struct_arg(e: s1) -> ti.i32:
+    ret = e.a + e.b.d
+    for i in range(3):
+        for j in range(2):
+            ret += e.b.c[i, j]
+    return ret
+
+arg = s1(a=1, b=s0(c=mat(1, 2, 3, 4, 5, 6), d=123))
+print(kernel_with_nested_struct_arg(arg))  # Prints 145
+```
```
[the remark this answers is not in the document]
Comment (lin-hitonami) [reply docs/lang/articles/kernels/kernel_function.md:129]: The code in the suggestion does not contain the argument and the kernel call. Could you please add them?

File: docs/lang/articles/kernels/kernel_function.md [hunk 3]
```diff
@@ -109,6 +109,19 @@ def my_kernel(x: int, y: float):
...
+```python
+transform_type = ti.types.struct(R=ti.math.mat3, T=ti.math.vec3)
+pos_type = ti.types.struct(x=ti.math.vec3, trans=transform_type)
+@ti.kernel
+def kernel_with_nested_struct_arg(p: pos_type) -> ti.math.vec3:
+    return p.trans.R @ p.x + p.trans.T
+trans = transform_type(ti.math.mat3(1), [1, 1, 1])
+p = pos_type(x=[1, 1, 1], trans=trans)
+print(kernel_with_nested_struct_arg(p))  # [4., 4., 4.]
+```
```
Comment (neozhaoliang) [inline docs/lang/articles/kernels/kernel_function.md:123]:
```suggestion
Here is another example of passing a nested struct argument with a matrix to a kernel by value:

```python
transform_type = ti.types.struct(R=ti.math.mat3, T=ti.math.vec3)
pos_type = ti.types.struct(x=ti.math.vec3, trans=transform_type)

@ti.kernel
def kernel_with_nested_struct_arg(p: pos_type) -> ti.math.vec3:
    return p.trans.R @ p.x + p.trans.T

trans = transform_type(ti.math.mat3(1), [1, 1, 1])
p = pos_type(x=[1, 1, 1], trans=trans)
print(kernel_with_nested_struct_arg(p))  # [4., 4., 4.]
```
```

Let's also add some new lines~

File: docs/lang/articles/kernels/kernel_function.md [hunk 4]
```diff
@@ -109,6 +109,20 @@ def my_kernel(x: int, y: float):
...
+pos_type = ti.types.struct(x=ti.math.vec3, trans=transform_type)
+
+@ti.kernel
+def kernel_with_nested_struct_arg(p: pos_type) -> ti.math.vec3:
+    return p.trans.R @ p.x + p.trans.T
+
+trans = transform_type(ti.math.mat3(1), [1, 1, 1])
+p = pos_type(x=[1, 1, 1], trans=trans)
+print(kernel_with_nested_struct_arg(p))  # [4., 4., 4.]
+
```
Comment (neozhaoliang) [inline docs/lang/articles/kernels/kernel_function.md:125]:
```suggestion
print(kernel_with_nested_struct_arg(p))  # [4., 4., 4.]
```
```

File: docs/lang/articles/kernels/kernel_function.md [hunk 5]
```diff
@@ -109,6 +109,21 @@ def my_kernel(x: int, y: float):
 my_kernel(1, 1.0)  # Prints 2.0
 ```
 
+Here is another example of passing a nested struct argument with a matrix to a kernel by value:
+
+```python
+transform_type = ti.types.struct(R=ti.math.mat3, T=ti.math.vec3)
+pos_type = ti.types.struct(x=ti.math.vec3, trans=transform_type)
```
Comment (jim19930609) [inline docs/lang/articles/kernels/kernel_function.md:116]: Shall we explain what `R`, `T`, `x`, `trans` are?

File: docs/lang/articles/kernels/kernel_function.md [hunk 6]
```diff
@@ -109,6 +109,21 @@ def my_kernel(x: int, y: float):
 my_kernel(1, 1.0)  # Prints 2.0
 ```
 
+Here is another example of passing a nested struct argument with a matrix to a kernel by value:
```
Comment (neozhaoliang) [inline docs/lang/articles/kernels/kernel_function.md:112]:
```suggestion
Here is another example of passing a nested struct argument with a matrix to a kernel by value, in which we created a struct type `transform_type` that contains two members: a rotation matrix `R` and a translation vector `T`. We then created another struct type `pos_type` that has `transform_type` as its member and passed an instance of `pos_type` to a kernel.
```

**Review by neozhaoliang — APPROVED** [review]
LGTM!

**Pull request merged**